rpc: Move the `generate` RPC call to rpcwallet

laanwj opened the pull request:
This makes it possible to mine to any wallet when multi-wallet mode is added. Solves the same problem as #10649, but IMO in a cleaner way.

It also gets rid of the circuitous `ScriptForMining` method on `CValidationInterface`, which really doesn't belong there.

After this change it's still possible to mine without wallet through `generatetoaddress`. I first proposed this in #7965.

laanwj committed:
rpc: Move the `generate` RPC call to rpcwallet

This makes it possible to mine to any wallet when multi-wallet mode is added.
Solves the same problem as #10649, but IMO in a cleaner way.

It also gets rid of the circuitous `ScriptForMining` method on
`CValidationInterface`, which really doesn't belong there.

After this change it's still possible to mine without wallet through
`generatetoaddress`.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
+            + HelpExampleCli("generate", "11")
+        );
+
+    int nGenerate = request.params[0].get_int();
```
nit: variable name should be lowercase, not hungarian

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
+        );
+
+    int nGenerate = request.params[0].get_int();
+    uint64_t nMaxTries = 1000000;
```
nit: variable name should be lowercase, not hungarian

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
+
+    int nGenerate = request.params[0].get_int();
+    uint64_t nMaxTries = 1000000;
+    if (request.params.size() > 1) {
```
nit: should handle case when `request.params[1]` is Null.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
+    pwallet->GetScriptForMining(coinbaseScript);
+
+    // If the keypool is exhausted, no script is returned at all.  Catch this.
+    if (!coinbaseScript)
```
nit: braces

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2922,6 +2923,47 @@ UniValue bumpfee(const JSONRPCRequest& request)
+        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+
+    //throw an error if no script was provided
+    if (coinbaseScript->reserveScript.empty())
```
nit: braces

jnewbery reviewed (commented):
tested ACK 8c410a8fbb5f1876484a3f10d25cf9c06a8ef875. This is a sensible change. Anything that moves us towards #7965 and away from circular libbitcoin_server.a/libbitcoin_wallet.a dependencies is a good thing in my book.

I realise that this is mostly code-move. There are a bunch of style nits in the moved code - entirely up to you if you want to address them as part of this PR or leave them as they are.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/10683/commits/8c410a8fbb5f1876484a3f10d25cf9c06a8ef875 though nit cleanup is great too.

jonasschnelli commented:
Nice and thanks for doing this:
utACK 8c410a8fbb5f1876484a3f10d25cf9c06a8ef875

laanwj commented:
Ok ok I'll add a commit to clean up the style nits, don't think it should be done at the time of the move as that just complicates review.

laanwj committed:
rpc: Update `generate` for developer notes

Fix nits by John Newbery.

jnewbery commented:
utACK 4cdb6209de2f712cd5847373b6bc2fb6401fb739. No more nits. Thanks!

> don't think it should be done at the time of the move as that just complicates review.

I'd agree if the nit fixes were in the same commit, but if they're done in their own commit either before or after the move it adds almost no review burden (as long as reviewers are reviewing commitwise)

laanwj commented:
> I'd agree if the nit fixes were in the same commit, but if they're done in their own commit either before or after the move it adds almost no review burden (as long as reviewers are reviewing commitwise)

Yes that's what I meant, with a separate commit it's fine.

TheBlueMatt commented on src/rpc/mining.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef BITCOIN_RPC_MINING_H
+#define BITCOIN_RPC_MINING_H
+
+#include <script/script.h>
```
Shouldnt this be a "-style include?

laanwj replied:
Yes, good point

TheBlueMatt reviewed (commented):
utACK 4cdb6209de2f712cd5847373b6bc2fb6401fb739, thanks for cleaning up CValidationInterface.

laanwj commented:
Updated `#include` style and squashed this into the first commit, where `mining.h` was introduced:
8c410a8fbb5f1876484a3f10d25cf9c06a8ef875 → df7e2f057b6c9f0f7c950f9077dc63a577f54117
4cdb6209de2f712cd5847373b6bc2fb6401fb739 → 2a962834febc9f56126ef06cf1bd5e1b02370278
